### ISSUE TITLE
No output or errors from nunit tests - 1.10.0-beta2

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit3-vs-adapter#329]
This issue is moved from the OmniSharp repository, 1.10.0-beta2 refers to the C# extension for Visual Studio Code which also uses the adapter.

_From @tomzo on May 14, 2017 9:54_

## Environment data
`dotnet --info` output:
```
.NET Command Line Tools (1.0.4)

Product Information:
 Version:            1.0.4
 Commit SHA-1 hash:  af1e6684fd

Runtime Environment:
 OS Name:     debian
 OS Version:  8
 OS Platform: Linux
 RID:         debian.8-x64
 Base Path:   /usr/share/dotnet/sdk/1.0.4
```
VS Code version:
C# Extension version: `1.12.1`

omnisharp-vscode: `1.10.0-beta2`

## Steps to reproduce

Here is a repo which exactly reproduces this https://github.com/tomzo/bug-nunit-no-output-vscode

Just run
```
dotnet restore
dotnet test DefaultTests/DefaultTests.csproj
```

Then try the same in vscode by clicking on `run test` on one of the tests.

## Expected  behavior

There are tests which are failingand there is standard output printed in one of the tests.
```csharp
[Test]
        public void Test1()
        {
			new Class1 ();
            Console.WriteLine("Test body running..");
        }

		[Test]
        public void Test2()
        {
            Assert.AreEqual(2,26);
        }
```
I would expect to see both the stdout and especially **failed asserts from nunit test cases**.

## Actual behavior

In VsCode  .NET Test Log output just counts the failed tests
```
NUnit Adapter 3.8.0.0: Test discovery starting
NUnit Adapter 3.8.0.0: Test discovery complete
NUnit Adapter 3.8.0.0: Test execution started
Running selected tests in /ide/work/nunit-repro/DefaultTests/bin/Debug/netcoreapp1.0/DefaultTests.dll
NUnit3TestExecutor converted 2 of 2 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete

Total tests: 1. Passed: 0. Failed: 1. Skipped: 0
```

The output when running with `dotnet` is fine:
```
$ dotnet test  DefaultTests/DefaultTests.csproj 
Build started, please wait...
/usr/share/dotnet/sdk/1.0.4/Sdks/Microsoft.NET.Sdk/Sdk/Sdk.targets(40,3): warning MSB4011: "/usr/share/dotnet/sdk/1.0.4/Microsoft.CSharp.targets" cannot be imported again. It was already imported at "/ide/work/nunit-repro/ExampleLib/ExampleLib.csproj (28,3)". This is most likely a build authoring error. This subsequent import will be ignored. 
Build completed.

Test run for /ide/work/nunit-repro/DefaultTests/bin/Debug/netcoreapp1.0/DefaultTests.dll(.NETCoreApp,Version=v1.0)
Microsoft (R) Test Execution Command Line Tool Version 15.0.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
NUnit Adapter 3.8.0.0: Test execution started
Running all tests in /ide/work/nunit-repro/DefaultTests/bin/Debug/netcoreapp1.0/DefaultTests.dll
NUnit3TestExecutor converted 2 of 2 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete
Failed   Test2
Error Message:
   Expected: 2
  But was:  26

Stack Trace:
at DefaultTests.UnitTest1.Test2() in /ide/work/nunit-repro/DefaultTests/UnitTest1.cs:line 20


Total tests: 2. Passed: 1. Failed: 1. Skipped: 0.
Test Run Failed.
Test execution time: 1.8351 Seconds
```

_Copied from original issue: OmniSharp/omnisharp-vscode#1480_

**Comment by rprouse** [comment]
_From @bernardbr on May 15, 2017 12:45_

Hello @tomzo!
Are you using the alpha version of NUnit?

```xml
  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.0.0"/>
    <PackageReference Include="NUnit" Version="3.6.1"/>
    <PackageReference Include="NUnit3TestAdapter" Version="3.8.0-alpha1"/>
  </ItemGroup>
```

**Comment by rprouse** [comment]
_From @tomzo on May 15, 2017 12:46_

Hi @bernardbr ,

Exactly, as you can see [here](https://github.com/tomzo/bug-nunit-no-output-vscode/blob/master/DefaultTests/DefaultTests.csproj#L28)

**Comment by rprouse** [comment]
_From @bernardbr on May 15, 2017 13:2_

Hello @tomzo!
I cloned your example and changed only .csproj and after this worked fine.  Could you check if it resolves your problem?

[bug-nunit-no-output-vscode.zip](https://github.com/OmniSharp/omnisharp-vscode/files/1001124/bug-nunit-no-output-vscode.zip)

**Comment by rprouse** [comment]
_From @bernardbr on May 15, 2017 13:5_

Oh, I also created the config files ```launch.json``` and ```tasks.json```

```powershell
Build started, please wait...
Build completed.

Test run for c:\Projetos\testes\bug-nunit-no-output-vscode\DefaultTests\bin\Debug\netcoreapp1.1\DefaultTests.dll(.NETCoreApp,Version=v1.1)
Ferramenta de Linha de Comando de Execução de Teste da Microsoft (R) Versão 15.0.0.0
Copyright (c) Microsoft Corporation. Todos os direitos reservados.

Iniciando execução de teste, espere...
NUnit Adapter 3.8.0.0: Test execution started
Running all tests in c:\Projetos\testes\bug-nunit-no-output-vscode\DefaultTests\bin\Debug\netcoreapp1.1\DefaultTests.dll
NUnit3TestExecutor converted 2 of 2 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete
Test2 com falha
Mensagem de erro:
   Expected: 2
  But was:  26

Rastreamento de pilha:
at DefaultTests.UnitTest1.Test2() in c:\Projetos\testes\bug-nunit-no-output-vscode\DefaultTests\UnitTest1.cs:line 22


Total de testes: 2. Aprovados: 1. Com falha: 1. Ignorados: 0.
Falha na Execução de Teste.
Tempo de execução de teste: 1,4842 Segundos
```

**Comment by rprouse** [comment]
_From @tomzo on May 15, 2017 13:19_

> I cloned your example and changed only .csproj and after this worked fine

It does not help for me, we must have something different in environment?

I downloaded your example. Ran `dotnet restore` then opened in vscode, and this is the output I see after clicking on `run test` above `Test2`:
```
NUnit Adapter 3.8.0.0: Test discovery starting
NUnit Adapter 3.8.0.0: Test discovery complete
NUnit Adapter 3.8.0.0: Test execution started
Running selected tests in /ide/work/nunit/DefaultTests/bin/Debug/netcoreapp1.1/DefaultTests.dll
NUnit3TestExecutor converted 2 of 2 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete

Total tests: 1. Passed: 0. Failed: 1. Skipped: 0
```

Other than simplifying the csproj, you have changed target framework to `netcoreapp1.1`, which was `netcoreapp1.0`, do you think this would have any impact? I am running all this on top of official docker image `microsoft/dotnet:1.0.5-sdk`

I will add you as collaborator to that repo. Please push to some branch, instead of the zips.

**Comment by rprouse** [comment]
_From @bernardbr on May 15, 2017 13:37_

@tomzo In my machine, your project won't works with ```DebugType``` tag on DefaultTests.csproj
I Just removed it and the tests ran!

I Pushed to https://github.com/tomzo/bug-nunit-no-output-vscode/tree/without-debug-type

**Comment by rprouse** [comment]
_From @tomzo on May 15, 2017 13:44_

> In my machine, your project won't works with DebugType tag on DefaultTests.csproj
I Just removed it and the tests ran!

Tests are running fine in my environment either with or without `DebugType`. But the issue is about the test output in vscode.
 `Test2` is failing because of ` Assert.AreEqual(2,26);`  and the **error is not shown**. Can you post your output from visual studio code when you run that test?

**Comment by rprouse** [comment]
_From @bernardbr on May 15, 2017 14:9_

But in my machine the output for test task works fine:

```powershell
Build started, please wait...
C:\Program Files\dotnet\sdk\1.0.4\Sdks\Microsoft.NET.Sdk\Sdk\Sdk.targets(40,3): warning MSB4011: "C:\Program Files\dotnet\sdk\1.0.4\Microsoft.CSharp.targets" cannot be imported again. It was already imported at "c:\Projetos\testes\bug-nunit-no-output-vscode\ExampleLib\ExampleLib.csproj (28,3)". This is most likely a build authoring error. This subsequent import will be ignored. 
Build completed.

Test run for c:\Projetos\testes\bug-nunit-no-output-vscode\DefaultTests\bin\Debug\netcoreapp1.0\DefaultTests.dll(.NETCoreApp,Version=v1.0)
Ferramenta de Linha de Comando de Execução de Teste da Microsoft (R) Versão 15.0.0.0
Copyright (c) Microsoft Corporation. Todos os direitos reservados.

Iniciando execução de teste, espere...
NUnit Adapter 3.8.0.0: Test execution started
Running all tests in c:\Projetos\testes\bug-nunit-no-output-vscode\DefaultTests\bin\Debug\netcoreapp1.0\DefaultTests.dll
NUnit3TestExecutor converted 2 of 2 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete
Test2 com falha
Mensagem de erro:
   Expected: 2
  But was:  26

Rastreamento de pilha:
at DefaultTests.UnitTest1.Test2() in c:\Projetos\testes\bug-nunit-no-output-vscode\DefaultTests\UnitTest1.cs:line 20

Total de testes: 2. Aprovados: 1. Com falha: 1. Ignorados: 0.
Falha na Execução de Teste.
Tempo de execução de teste: 2,5723 Segundos
```

Let's compare our enviroment settings:
#### Dotnet info: 
.NET Command Line Tools (1.0.4)

Product Information:
 Version:            1.0.4
 Commit SHA-1 hash:  af1e6684fd

Runtime Environment:
 OS Name:     Windows
 OS Version:  6.1.7601
 OS Platform: Windows
 RID:         win7-x64
 Base Path:   C:\Program Files\dotnet\sdk\1.0.4

#### VsCode version:
Version 1.12.1
Commit f6868fce3eeb16663840eb82123369dec6077a9b
Date 2017-05-04T21:26:50.689Z
Shell 1.6.6
Renderer 56.0.2924.87
Node 7.4.0


#### Omnisharp-vscode version:
1.10.0-beta2

#### Nunit version:
3.8.0-alpha1

In my case I'm running on Windows. I'll try run on Ubuntu.


PS: With ```DebugType``` tag the test didn't work.

**Comment by rprouse** [comment]
_From @tomzo on May 15, 2017 14:21_

@bernardbr thank you for helping with this. But I think that you keep invoking a different action.
When I **run a test task** like you just did, then I also get the correct output:
```
Starting test execution, please wait...
NUnit Adapter 3.8.0.0: Test execution started
Running all tests in /ide/work/nunit/bug-nunit-no-output-vscode/DefaultTests/bin/Debug/netcoreapp1.0/DefaultTests.dll
NUnit3TestExecutor converted 2 of 2 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete
Failed   Test2
Error Message:
   Expected: 2
  But was:  26

Stack Trace:
at DefaultTests.UnitTest1.Test2() in /ide/work/nunit/bug-nunit-no-output-vscode/DefaultTests/UnitTest1.cs:line 20
Total tests: 2. Passed: 1. Failed: 1. Skipped: 0.
Test Run Failed.
Test execution time: 1.8279 Seconds
```
But that **runs all tests**. 

There is no output however when I **click on `run test`** above the test name. Like on this screenshot
![image](https://cloud.githubusercontent.com/assets/116161/24731864/c616b062-1a21-11e7-9f9b-76387b6d4890.png)

**Comment by rprouse** [comment]
_From @bernardbr on May 15, 2017 14:37_

Ooh! Ok. I'm sorry, I didn't understand it before!

I don't know, but for me, in this case, the nunit does the "focus" only in "ran test" ignoring the other ones. You can try run the ```Test1``` and see this output:

```powershell
Running test DefaultTests.UnitTest1.Test1...
Microsoft (R) Build Engine version 15.1.1012.6693
Copyright (C) Microsoft Corporation. All rights reserved.

C:\Program Files\dotnet\sdk\1.0.4\Sdks\Microsoft.NET.Sdk\Sdk\Sdk.targets(40,3): warning MSB4011: "C:\Program Files\dotnet\sdk\1.0.4\Microsoft.CSharp.targets" cannot be imported again. It was already imported at "c:\Projetos\testes\bug-nunit-no-output-vscode\ExampleLib\ExampleLib.csproj (28,3)". This is most likely a build authoring error. This subsequent import will be ignored. 
  ExampleLib -> c:\Projetos\testes\bug-nunit-no-output-vscode\ExampleLib\bin\Debug\netcoreapp1.0\ExampleLib.dll
  DefaultTests -> c:\Projetos\testes\bug-nunit-no-output-vscode\DefaultTests\bin\Debug\netcoreapp1.0\DefaultTests.dll

Build succeeded.

C:\Program Files\dotnet\sdk\1.0.4\Sdks\Microsoft.NET.Sdk\Sdk\Sdk.targets(40,3): warning MSB4011: "C:\Program Files\dotnet\sdk\1.0.4\Microsoft.CSharp.targets" cannot be imported again. It was already imported at "c:\Projetos\testes\bug-nunit-no-output-vscode\ExampleLib\ExampleLib.csproj (28,3)". This is most likely a build authoring error. This subsequent import will be ignored. 
    1 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.31


NUnit Adapter 3.8.0.0: Test discovery starting
NUnit Adapter 3.8.0.0: Test discovery complete
NUnit Adapter 3.8.0.0: Test execution started
Running selected tests in c:\Projetos\testes\bug-nunit-no-output-vscode\DefaultTests\bin\Debug\netcoreapp1.0\DefaultTests.dll
NUnit3TestExecutor converted 2 of 2 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete

Total tests: 1. Passed: 1. Failed: 0. Skipped: 0
```

I miss an option to run and debug all test fixtures, but I'm afraid this doesn't exist on VsCode.

Maybe @DustinCampbell and @rprouse could help us to clarify it!

**Comment by rprouse** [comment]
_From @tomzo on May 15, 2017 14:50_

> I miss an option to run and debug all test fixtures, but I'm afraid this doesn't exist on VsCode.

That's not a big problem.

> You can try run the Test1 and see this output:

Yes, I can see that output from the very begging.
To be clear - my point is that for `Test2` it should contain Error message and stack trace:
```
NUnit3TestExecutor converted 2 of 2 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete
Failed   Test2
Error Message:
   Expected: 2
  But was:  26

Stack Trace:
at DefaultTests.UnitTest1.Test2() in /ide/work/nunit/bug-nunit-no-output-vscode/DefaultTests/UnitTest1.cs:line 20
```
But now it is just
```
NUnit Adapter 3.8.0.0: Test discovery starting
NUnit Adapter 3.8.0.0: Test discovery complete
NUnit Adapter 3.8.0.0: Test execution started
Running selected tests in /ide/work/nunit/DefaultTests/bin/Debug/netcoreapp1.1/DefaultTests.dll
NUnit3TestExecutor converted 2 of 2 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete

Total tests: 1. Passed: 0. Failed: 1. Skipped: 0
```

**Comment by rprouse** [comment]
_From @bernardbr on May 15, 2017 15:12_

I agree with you!

I miss it too... Always when a single test fails, I need run all tests to get a stack trace...

But I don't know if it is a fail of OmniSharp or a limitation of NUnit test adapter!

**Comment by rprouse** [comment]
Looks like this is an NUnit issue. The NUnit adapter has only ever been used in Visual Studio before. In Visual Studio the errors are displayed in a test results window, not in the output. I created an xUnit project and they output the results to the output so we should do the same.

**Comment by rprouse** [comment]
I checked xUnit in Visual Studio and it outputs errors to the output window there too, so it isn't special behavior for Visual Studio Code. It was likely added for the `dotnet test` command so that errors appear on the command line. We should do the same.

#### xUnit's output for a failing test,

```
------ Run test started ------
[xUnit.net 00:00:00.4649832]   Starting:    XUnitTestProject
[xUnit.net 00:00:00.6220831]     XUnitTestProject.UnitTest1.Test1 [FAIL]
[xUnit.net 00:00:00.6234021]       Assert.Equal() Failure
[xUnit.net 00:00:00.6235210]       Expected: 1
[xUnit.net 00:00:00.6235536]       Actual:   42
[xUnit.net 00:00:00.6244108]       Stack Trace:
[xUnit.net 00:00:00.6254159]         c:\src\UnitTestProject1\XUnitTestProject\UnitTest1.cs(11,0): at XUnitTestProject.UnitTest1.Test1()
[xUnit.net 00:00:00.6398506]   Finished:    XUnitTestProject
========== Run test finished: 1 run (0:00:01.1940726) ==========
```

#### NUnit's output for a failing test

```
------ Run test started ------
NUnit Adapter 3.8.0.0: Test execution started
Running selected tests in c:\src\UnitTestProject1\NUnitTestProject\bin\Debug\netcoreapp1.1\NUnitTestProject.dll
NUnit3TestExecutor converted 1 of 1 NUnit test cases
NUnit Adapter 3.8.0.0: Test execution complete
========== Run test finished: 1 run (0:00:00.9459785) ==========
```

**Comment by CharliePoole** [comment]
@rprouse What about running in the IDE? We have always given VS the output to display, which it does quite nicely, and avoided cluttering the output window. How does xUnit display in that case?

We do have one bit of code in the adapter where it detects it's running in the IDE and acts differently. If necessary, we can expand that usage.

**Comment by rprouse** [comment]
The above output is from running in the IDE. xUnit outputs errors to the output window. If we can tell we are running under Visual Studio, we could suppress the output, but we need to make sure it is Visual Studio, not some other IDE like Visual Studio Code that does not have the same capabilities.

**Comment by CharliePoole** [comment]
That raises an interesting question. Does our existing logic that detects the IDE work correctly under Code? I'm on my phone now, but I'll take a look later.

**Comment by CharliePoole** [comment]
Here's where we do it... from TestConverter.cs...

```C#
                string message = node?.SelectSingleNode("message")?.InnerText;
                // If we're running in the IDE, remove any caret line from the message
                // since it will be displayed using a variable font and won't make sense.
                if (!string.IsNullOrEmpty(message) && NUnitTestAdapter.IsRunningUnderIDE)
                {
                    string pattern = NL + "  -*\\^" + NL;
                    message = Regex.Replace(message, pattern, NL, RegexOptions.Multiline);
                }
```

`IsRunningUnderIDE` is true when the entry program is  "vstest.executionengine",  "vstest.discoveryengine" or "TE.ProcessHost".

Should be easy enough to check by creating a failing equal assert and seeing if the error message includes the caret line.

**Comment by sebwills** [comment]
With reference to @CharliePoole 's previous comment: my experience (VS Code on a Mac) with a test containing `Assert.That("12345", Is.EqualTo("12355"));` is that the 'run test' link above the test method produces no useful output other than a summary line saying "Total Tests: 1, ... Failed: 1"  (i.e. it doesn't include the assertion failure message at all, let alone with a caret line). 

If I run `dotnet test` in the VS Code integrated terminal, I get the full error including a caret line, as expected.

So it seems like something more than `IsRunningUnderIDE` is at play here?

**Comment by CharliePoole** [comment]
@sebwills Are you sure you are running the adapter in that situation? I'm dubious, since it isn't written for use with vscode, which I believe has its own interfaces. I'd love it if that were possible!

**Comment by sebwills** [comment]
@CharliePoole I may be mistaken but it looks that way to me. The output when I click on the 'run test' link is:

> NUnit Adapter 3.9.0.0: Test discovery starting
> NUnit Adapter 3.9.0.0: Test discovery complete
> NUnit Adapter 3.9.0.0: Test execution started
> Running selected tests in /Users/.........X.dll
> NUnit3TestExecutor converted 40 of 40 NUnit test cases
> NUnit Adapter 3.9.0.0: Test execution complete
> 
> Total tests: 1. Passed: 0. Failed: 1. Skipped: 0

Isn't 'NUnit Adaptor 3.9.0.0' "the adapter" ?

**Comment by CharliePoole** [comment]
That's the adapter. And you are running vscode, not visual studio? If so, it's something I'm not aware of.

**Comment by sebwills** [comment]
@CharliePoole Yes, I'm running vscode on macOS.

**Comment by CharliePoole** [comment]
OK I get it now! Vscode has an extension that runs dotnet test. That ends up bringing in our adapter. I think the problem may be that our development has not focused on this particular use of the adapter. 

@OsirisTerje Do you have any more info on this? I'm thinking of writing an adapter for use in vscode, so the question is of interest to me.

**Comment by OsirisTerje** [comment]
@CharliePoole That is the way I have been using VS Code, I run the terminal window, and run dotnet test from there.  It is running the adapter, so all fine there.
However, I haven't really used the Run Test and Debug Test links in the code.  It see it starts up the dotnet test in the terminal window.  I am pretty sure it was workign earlier, but right now I can't make it work for either MSTest nor NUnit.  It runs the adapter, but it don't find any tests.  I have to check more to see if I have everything correctly set up. 
On whether it needs another adapter, I dont think so.  This is the way it works, afaik, for all test frameworks.  If there was a more integrated approach, it could be interesting though, but for me it has been working good enough.  
I am mostly using VS code for smaller projects, and for Python coding.

**Comment by CharliePoole** [comment]
@OsirisTerje Thanks for the info!

**Comment by rprouse** [comment]
@CharliePoole and @OsirisTerje I did do some work with Dustin Campbell who works on the C# extension for VSCode to ensure that our adapter works there, especially with CodeLens. It works, but we haven't done much testing and as far as I know, none of the .NET test adapters work very well. I still haven't managed to get a very integrated experience for C# development in VSCode, unlike it's stellar support for Typescript.

**Comment by OsirisTerje** [comment]
@rprouse When I tested it a couple of days back, it didn't work with codelens in VSCode.  Did you do anything to make it work?

**Comment by rprouse** [comment]
I just tested it tonight with the following steps.

If you haven't already installed our .NET Core templates, do so,

```
dotnet new -i NUnit3.DotNetNew.Template
```

Then create a new test project,

```
mkdir nunit-test
cd nunit-test
dotnet new nunit
```

Open the project in VSCode,

```
code .
```

Displays a warning about required assets being missing. Install them,

![image](https://user-images.githubusercontent.com/493828/35366966-f9ce00b2-0149-11e8-8277-7b68dbf55f82.png)

Open `UnitTest1.cs`, put a breakpoint on the `Assert.Pass` and click debug test. Tests are launched and breakpoint is hit.

![image](https://user-images.githubusercontent.com/493828/35367025-3b185298-014a-11e8-8041-1aeea6540213.png)

All this assumes you have .NET Core 2.0 installed along with the C# extension for VSCode.

![image](https://user-images.githubusercontent.com/493828/35367064-74f80062-014a-11e8-8b93-2b4c505adc3f.png)

**Comment by OsirisTerje** [comment]
@rprouse  I can repro the same here, that there is no console output when using the codelens "Run Tests".  Did you get any more info from MS/Mr.Campbell about this?  Is this something on their or our side?

**Comment by rprouse** [comment]
No console output using CodeLens is probably on the C# extension side. I haven't followed up on their side since this issue was initially filed though.

**Comment by DustinCampbell** [comment]
@OsirisTerje: Where did you raise the issue? I don't see it on the C# for VS Code repo.

**Comment by OsirisTerje** [comment]
@DustinCampbell I have raised it internally with the product group.

**Comment by OsirisTerje** [comment]
@DustinCampbell  The issue is tracked here: https://github.com/OmniSharp/omnisharp-vscode/issues/2268

**Comment by DustinCampbell** [comment]
Thanks @OsirisTerje!

**Comment by OsirisTerje** [comment]
Fixed in OmniSharp/omnisharp-vscode#2268
